### PR TITLE
New Anomaly Dashboard

### PR DESCRIPTION
**Pull request by npawar**
This pull request contains CRUD in AnomalyResouce for anomalies, functions and reports. It also contains a new dashboard interface for the anomalies.

### CODE REVIEW

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/common/ThirdEyeConfiguration.java [hunk 1]
```diff
@@ -82,4 +88,32 @@ public void setBlacklistCollections(String blacklistCollections) {
...
 
+  public String getFunctionConfigPath() {
+    return getRootDir() + "/detector-config/anomaly-functions/functions.properties";
+  }
+
+  public FailureEmailConfiguration getFailureEmailConfig() {
+    return failureEmailConfig;
+  }
+
+  public void setFailureEmailConfig(FailureEmailConfiguration failureEmailConfig) {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/common/ThirdEyeConfiguration.java]: Remove unneeded method

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/DetectorHttpUtils.java [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    } finally {
+      if (res.getEntity() != null) {
+        EntityUtils.consume(res.getEntity());
+      }
+    }
+    return response;
+  }
+
+  public static void main(String[] args) throws ClientProtocolException, IOException {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/DetectorHttpUtils.java]: delete main method

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/ThirdEyeDetectorApplication.java [hunk 3]
```diff
@@ -240,8 +229,26 @@ public void serverStarted(Server server) {
...
+                        LOG.info("Starting {}", config);
+                        emailReportJobManager.start(config.getId());
+                      } catch (Exception e) {
+                        LOG.error("Failed to schedule report " + config.getId(), e);
+                        failedToStart.add(config);
+                      }
+                    }
+                  }
+                  if (!failedToStart.isEmpty()) {
+                    LOG.warn("{} functions failed to start!: {}", failedToStart.size(), failedToStart);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/ThirdEyeDetectorApplication.java]: s/functions/reports/

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java [hunk 4]
```diff
@@ -1,26 +1,32 @@
...
 
+import java.util.ArrayList;
 import java.util.List;
 
 import org.hibernate.SessionFactory;
 
 import com.linkedin.thirdeye.detector.api.AnomalyFunctionSpec;
 
 import io.dropwizard.hibernate.AbstractDAO;
+import io.dropwizard.hibernate.UnitOfWork;
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java]: unneeded import?

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/EmailConfigurationDAO.java [hunk 5]
```diff
@@ -14,12 +14,20 @@ public EmailConfigurationDAO(SessionFactory sessionFactory) {
...
 
-  public Long create(EmailConfiguration emailConfiguration) {
-    return persist(emailConfiguration).getId();
+  public Long createOrUpdate(EmailConfiguration emailConfiguration) {
+    long id = persist(emailConfiguration).getId();
+    currentSession().getTransaction().commit();
+    return id;
   }
 
   public void delete(Long id) {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/EmailConfigurationDAO.java]: delete if this is no longer being used.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java [hunk 6]
```diff
@@ -40,4 +50,15 @@ public void delete(Long id) {
     return list(namedQuery("com.linkedin.thirdeye.api.AnomalyFunctionSpec#findAllByCollection")
         .setParameter("collection", collection));
   }
+
+  public List<String> findDistinctMetricsByCollection(String collection) {
+    List<String> metrics = new ArrayList<>();
+    List<AnomalyFunctionSpec> anomalyFunctionSpecs = list(namedQuery("com.linkedin.thirdeye.api.AnomalyFunctionSpec#findDistinctMetricsByCollection")
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java]: rename variable, add comment explaining that these are actually strings

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java [hunk 7]
```diff
@@ -32,6 +38,10 @@ public void delete(Long id) {
     currentSession().delete(anomalyFunctionSpec);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/db/AnomalyFunctionSpecDAO.java]: delete if no longer being used

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/driver/TestAnomalyApplication.java [hunk 8]
```diff
@@ -96,7 +96,6 @@ public void run(ThirdEyeDetectorConfiguration config, Environment environment) t
     final ThirdEyeClient thirdEyeClient =
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/detector/driver/TestAnomalyApplication.java]: go ahead and delete this whole class

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/util/ThirdEyeUtils.java [hunk 9]
```diff
@@ -148,4 +150,21 @@ public static String getAliasFromCollection(String collection) throws ExecutionE
     return result;
   }
 
+  public static String constructCron(String scheduleMinute, String scheduleHour, String repeatEvery) {
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/util/ThirdEyeUtils.java]: `repeatEvery` should be a TimeUnit.
validation on scheduleMinute and scheduleHour
Add unit test(s)

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 10]
```diff
@@ -0,0 +1,608 @@
...
+import com.linkedin.thirdeye.detector.db.EmailConfigurationDAO;
+import com.linkedin.thirdeye.util.ThirdEyeUtils;
+
+@Path(value = "/dashboard")
+@Produces(MediaType.APPLICATION_JSON)
+public class AnomalyResource {
+  private static final ThirdEyeCacheRegistry CACHE_REGISTRY_INSTANCE = ThirdEyeCacheRegistry.getInstance();
+  private static final Logger LOG = LoggerFactory.getLogger(AnomalyResource.class);
+
+  private static final String DEFAULT_SMTP_PORT = "25";
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: move smtp port to configs

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 11]
```diff
@@ -0,0 +1,608 @@
...
+  public List<String> viewMetricsForDataset(@QueryParam("dataset") String dataset) {
+    List<String> metrics = anomalyFunctionSpecDAO.findDistinctMetricsByCollection(dataset);
+    return metrics;
+  }
+
+  // View anomalies for collection
+  @GET
+  @UnitOfWork
+  @Path("/anomalies/view")
+  public List<AnomalyResult> viewAnomaliesInRange(@QueryParam("dataset") String dataset,
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: check if there's a 'required' flag for dataset. If not, add validation.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 12]
```diff
@@ -0,0 +1,608 @@
...
+      @QueryParam("windowSize") String windowSize,
+      @QueryParam("windowUnit") String windowUnit,
+      @QueryParam("windowDelay") String windowDelay,
+      @QueryParam("windowDelayUnit") String windowDelayUnit,
+      @QueryParam("scheduleMinute") String scheduleMinute,
+      @QueryParam("scheduleHour") String scheduleHour,
+      @QueryParam("repeatEvery") String repeatEvery,
+      @QueryParam("exploreDimension") String exploreDimensions,
+      @QueryParam("properties") String properties,
+      @QueryParam("isActive") boolean isActive)
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: add comment explaining that bucketSize, bucketUnit are defaulted to collection granularity.
Add filters param.

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 13]
```diff
@@ -0,0 +1,608 @@
...
+    }
+    // call endpoint to stop if active
+    if (anomalyFunctionSpec.getIsActive()) {
+      detectorHttpUtils.disableAnomalyFunction(String.valueOf(id));
+    }
+
+    CollectionSchema schema = CACHE_REGISTRY_INSTANCE.getCollectionSchemaCache().get(dataset);
+    TimeGranularity dataGranularity = schema.getTime().getDataGranularity();
+
+    anomalyFunctionSpec.setIsActive(false);
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: refactor shared code between create + update

File: thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java [hunk 14]
```diff
@@ -0,0 +1,608 @@
...
+  @POST
+  @UnitOfWork
+  @Path("/anomaly-function/adhoc")
+  public Response runAdhocAnomalyFunctions(@QueryParam("id") Long id, @QueryParam("functionName") String functionName,
+      @QueryParam("windowStartIso") String windowStartIso, @QueryParam("windowEndIso") String windowEndIso)
+          throws Exception {
+
+    AnomalyFunctionSpec anomalyFunctionSpec = anomalyFunctionSpecDAO.findById(id);
+    if (StringUtils.isEmpty(windowStartIso) || StringUtils.isEmpty(windowEndIso)) {
+      int windowSize = anomalyFunctionSpec.getWindowSize();
```
Comment (jiateoh) [inline thirdeye/thirdeye-pinot/src/main/java/com/linkedin/thirdeye/dashboard/resources/AnomalyResource.java]: no need to do now()-based logic here, just construct url via httpl util and pass in the query params for start/end if they are present.

**Comment by jiateoh** [comment]
LGTM after fixing comments

**Comment by npawar** [comment]
Fixed comments

**Comment by npawar** [comment]
Testing done: Tested all endpoints using curl calls
Anomalies: 
1) View - Tested that all anomalies that match filters are returned
Anomaly functions:
1) Create - Tested that anomaly function is created successfully in dashboard, and if it is active, it is also started in the scheduler
2) Update - Tested that the anomaly function is updated successfully, and the scheduler is notified of the changes in function description or active/inactive
3) View - Tested that all anomaly functions that match filter are displayed
4) Delete - Tested that the anomaly function is deleted from the database, and the scheduler is notified to remove the schedule
Email Configs:
1) Create - Tested that email config is created successfully in dashboard, and if it is active, it is also started in the scheduler
2) Update - Tested that the email config is updated successfully, and the scheduler is notified of the changes in function description or active/inactive
3) View - Tested that all email config that match filter are displayed
4) Delete - Tested that the email config is deleted from the database, and the scheduler is notified to remove the schedule

**Pull request merged**